README.md
# CSharp-like-properties-in-python.  
### Properties in python that looks a bit like C# ones   
### These can be though of as an alternative implementation of the @property decorator.
## Pros:
### 1. A bit less of typing.
### 2. Better code structuring.
### 3. More error-prone, as getter/setter/deleter names are predetermined. 
## Cons:
### 1. The implementation can miss something that can be done with good old property declaration.
### 2. A lot of things works "behind the scene", i.e. implicitly.

##### functions get_all_nested and free_var are borrowed from repository at https://github.com/sansyrox/nestd  
##### with many thanks to sansyrox 
   
##### To use this code copy the file cslike_props.py in your project.   
examle of usage:  

```python

from cslike_props import cprop, pprop, _del_cprop_attr

@cprop
class TestClass:

    def __init__(self, test_prop_one, test_prop_py, test_prop_two):
        self.test_prop_one = test_prop_one
        self.test_prop_two = test_prop_two
        self.test_prop_py = test_prop_py

    def funct(self):
        print('in TestClass:funct')

    # if uncommented there will be AttributeError: attribute test_prop_one_get_v already exists
    # def test_prop_one_get_v(self):
    #     pass

    @pprop
    def test_prop_one(self):
        # property name will be the same as the function name - test_prop_one
        # and the member variable _test_prop_one will be added to the class
        # and the list of all names of all the added attributes for the property
        # will be added to a getter function test_prop_one_get_v as the attribute 
        # with the name attr_names
        
        # the getter declaration, underlying getter name is test_prop_one_get_v,
        # the function test_prop_one_get_v will be added to a definition of this class
        def get_v(self):
            return self._test_prop_one * 5

        # the setter declaration, underlying setter name is test_prop_one_set_v
        # the function test_prop_one_set_v will be added to a definition of this class
        def set_v(self, value):
            self._test_prop_one = value * 5

        # the deleter declaration, underlying deleter name is test_prop_one_del_v
        # the function test_prop_one_del_v will be added to a definition of this class        
        def del_v(self):
            # the function _del_cprop_attr will be added to a definition of this class
            self._del_cprop_attr(self.test_prop_one_get_v.attr_names)

        # if uncommented there will be
        # ValueError: test_prop_one property should define no more than three inner functions
        # def delete2(self):
        #     pass

    # example of pure pythonic property for comparison with
    # example of the C# - like property test_prop_one property
    # As it can be seen here is a bit more code than in the test_prop_one example,
    # the functions in the property definition could be spread over the entire class
    # without any consequences
    # whereas the C# - like property requires
    # all the functions in the property definition are nested inside
    # property definition ->

    def test_prop_py_get(self):
        return self._test_prop_py * 5 ** 3

    def test_prop_py_set(self, value):
        self._test_prop_py = value * 5 ** 3

    def test_prop_py_delete(self):
        attrs = ['test_prop_py_get', 'test_prop_py_set', 'test_prop_py_delete']
        _del_cprop_attr(self, attrs)
        del self._test_prop_py

    test_prop_py = property(test_prop_py_get, test_prop_py_set, test_prop_py_delete)

    # <- example of pure pythonic property for comparison with
    # example of the C# - like property test_prop_one property

    @pprop
    def test_prop_two(self):
        # attribute name will be the same as the function name - test_prop_two

        def get_v(self):
            return self._test_prop_two * 5 ** 2

        def set_v(self, value):
            self._test_prop_two = value * 5 ** 2

x, y, z = 555, 5, 55555


t = TestClass(test_prop_one=x, test_prop_py=y, test_prop_two=z)

assert t.test_prop_one == x * 5 ** 2
assert t.test_prop_py == y * 5 ** 3 * 5 ** 3
assert t.test_prop_two == z * 5 ** 2 * 5 ** 2

a = 25
t.test_prop_one = a
b = 5
t.test_prop_two = b

assert t._test_prop_one == a * 5
assert t._test_prop_two == b * 5 ** 2

assert t.test_prop_one == a * 5 * 5
assert t.test_prop_two == b * 5 ** 2 * 5 ** 2


attr_one_names = ['_test_prop_one', 'test_prop_one_get_v', 'test_prop_one_set_v', 'test_prop_one']
assert t.test_prop_one_get_v.attr_names == \
       attr_one_names

for a in attr_one_names:
    assert hasattr(TestClass, a)

del t.test_prop_one

for a in attr_one_names:
    assert not hasattr(TestClass, a)

assert t.test_prop_two_get_v.attr_names == \
       ['_test_prop_two', 'test_prop_two_get_v', 'test_prop_two_set_v', 'test_prop_two']

attr_py_names = ['test_prop_py_get', 'test_prop_py_set', 'test_prop_py_delete']

for a in attr_py_names:
    assert hasattr(TestClass, a)

del t.test_prop_py

for a in attr_py_names:
    assert not hasattr(TestClass, a)

assert not hasattr(TestClass, '_test_prop_py')


```


history.md
2023/01/22 15:30 GMT +7 Version 0.01 submitted


cslike_props.py
import inspect
import types
from dataclasses import dataclass
from typing import Callable, Any, Type, Final, Set, List, Tuple, ClassVar, Dict, Optional

# https://github.com/sansyrox/nestd
# functions get_all_nested and free_var are borrowed from repository at https://github.com/sansyrox/nestd
# with many thanks to sansyrox


def free_var(val):
    """A function that wraps free variables."""

    def nested():
        return val

    return nested.__closure__[0]


def get_all_nested(fx, *context_vars):
    """Return all nested functions of a function.

    Arguments:
        fx (function or method): A function object with inner function(s).
        *context_vars: context variables corressponding inner functions in the order of occurence.

    Returns:
        A list of tuples, with the first element as the function name and the second element as function object.
        e.g. [("inner_function", <class function....>), ....]
    """
    if not isinstance(fx, (types.FunctionType, types.MethodType)):
        raise Exception("Supplied param is not a function or a method type")

    fx = fx.__code__
    context_variables = list(context_vars)

    output = []
    for const in fx.co_consts:
        if isinstance(const, types.CodeType):
            context = tuple(
                free_var(context_variables[0]) for name in const.co_freevars
            )
            context_variables = context_variables[1:]
            output.append(
                (
                    const.co_name,
                    types.FunctionType(const, globals(), None, None, context),
                )
            )

    return output


@dataclass
class PropVal:
    get_f_name: str = ''
    get_f_func: Optional[Callable[[], Any]] = None
    set_f_name: str = ''
    set_f_func: Optional[Callable[[Any], None]] = None
    del_f_name: str = ''
    del_f_func: Optional[Callable[[], None]] = None

    prop_name: str = ''

    _good: ClassVar[List[str]] = ['get_v', 'set_v', 'del_v']


def _del_cprop_attr(self: Type, attrs: List[str]):
    # delete attributes from a class
    klass = type(self)

    for a in attrs:
        delattr(klass, a)
        d = dir(self)
        if a in d:
            delattr(self, a)


# decorator to decorate class enabled for use C# - like properties
def cprop(klass: Type, arg_1=5): # Class c-Sharp - like decorator of PRoperty Void (without parametes)
    props: List[PropVal] = []
    good = PropVal._good  # ['get_v', 'set_v', 'del_v']
    good_names: Final[Set[str]] = set(good)
    # find all functions decorated with 'pprop' decorator:
    for k, f in klass.__dict__.items():
        if inspect.isfunction(f) and hasattr(f, 'ppropdecor'):
            inner_functions = get_all_nested(f, [])
            all_inner_functions = dict(inner_functions)

            if len(all_inner_functions) == 0:
                raise ValueError(f'{k} property should define at least one inner function')

            if len(all_inner_functions) > 3:
                raise ValueError(f'{k} property should define no more than three inner functions')

            banned = [k for k in all_inner_functions.keys() if k not in good_names]
            if len(banned) > 0:
                raise ValueError((f'{k} property has not acceptable item(s): {banned}, '
                                 f'acceptable items are: {good_names}'))

            if not good[0] in all_inner_functions.keys():
                raise ValueError(f'{k} property must have function {good[0]}')

            if not good[1] in all_inner_functions.keys() and good[2] in all_inner_functions.keys():
                raise ValueError((f'{k} property to have function {good[2]} ' 
                                  f'must have function {good[1]}'))
            props.append(PropVal(get_f_name=good[0],
                                 get_f_func=all_inner_functions[good[0]],
                                 set_f_name=(set_name := (good[1] if good[1] in all_inner_functions.keys()
                                                          else '')),
                                 set_f_func=(all_inner_functions[good[1]] if set_name else None),
                                 del_f_name=(del_name := (good[2] if good[2]
                                                          in all_inner_functions.keys() else '')),
                                 del_f_func=(all_inner_functions[good[2]] if del_name else None),
                                 prop_name=k))

    existing_keys = klass.__dict__.keys()

    def check_name(attribute_name: str):
        if attribute_name in existing_keys:
            raise AttributeError(f'attribute {attribute_name} already exists')

    delchecked: bool = False
    for p in props:
        attr_names = []

        def set_attrib(aname: str, value: Any) -> None:
            check_name(aname)
            setattr(klass, aname, value)
            attr_names.append(aname)
        delattr(klass, p.prop_name)

        attr_name = '_' + p.prop_name
        set_attrib(attr_name, None)
        prop_funcs = [p.get_f_func]
        get_func_name = p.prop_name + '_' + p.get_f_name
        set_attrib(get_func_name,  p.get_f_func)
        if p.set_f_func:
            prop_funcs.append(p.set_f_func)
            func_name = p.prop_name + '_' + p.set_f_name
            set_attrib(func_name, p.set_f_func)

            if p.del_f_func:
                prop_funcs.append(p.del_f_func)
                func_name = p.prop_name + '_' + p.del_f_name
                setattr(klass, func_name, p.del_f_func)

        # finally set the property itself
        check_name(p.prop_name)
        setattr(klass, p.prop_name, property(*prop_funcs))
        attr_names.append(p.prop_name)

        # set the list of names of attributes
        getattr(klass, get_func_name).attr_names = attr_names

        # set the function _del_cprop_attr as class method
        if not delchecked:
            delchecked = True
            aname = '_del_cprop_attr'
            check_name(aname)
            setattr(klass, aname, _del_cprop_attr)
    return klass


# # decorator to decorate C# - like properties
def pprop(fn, arg_1=5) -> Callable[..., Any]:  # property c-Sharp - like decorator of PRoperty Void (without parametes)
    fn.ppropdecor = True
    return fn


__version__ = 0.01

if __name__ == '__main__':
    # some tests
    @cprop
    class TestClass:

        def __init__(self, test_prop_one, test_prop_two):
            self.test_prop_one = test_prop_one
            self.test_prop_two = test_prop_two

        def funct(self):
            print('in TestClass:funct')

        # if uncommented there will be AttributeError: attribute test_prop_one_get_v already exists
        # def test_prop_one_get_v(self):
        #     pass

        @pprop
        def test_prop_one(self):
            # attribute name will be the same as the function name - test_prop_one

            def get_v(self):
                return self._test_prop_one

            def set_v(self, value):
                self._test_prop_one = value

            def del_v(self):
                self._del_cprop_attr(self.test_prop_one_get_v.attr_names)

            # if uncommented there will be
            # ValueError: test_prop_one property should define no more than three inner functions
            # def delete2(self):
            #     pass

        @pprop
        def test_prop_two(self):
            # attribute name will be the same as the function name - test_prop_two

            def get_v(self):
                return self._test_prop_two

            def set_v(self, value):
                self._test_prop_two = value


    t = TestClass(test_prop_two=555, test_prop_one=55555)
    t.test_prop_one = 5 * 5

    t.test_prop_two = 5

    assert t.test_prop_two == 5
    assert t.test_prop_one == 5 * 5

    attr_one_names = ['_test_prop_one', 'test_prop_one_get_v', 'test_prop_one_set_v', 'test_prop_one']
    assert t.test_prop_one_get_v.attr_names == \
           attr_one_names

    for a in attr_one_names:
        assert hasattr(TestClass, a) #  or hasattr(t, a)

    del t.test_prop_one

    for a in attr_one_names:
        assert (not hasattr(TestClass, a))

    assert t.test_prop_two_get_v.attr_names == \
           ['_test_prop_two', 'test_prop_two_get_v', 'test_prop_two_set_v', 'test_prop_two']
